### ISSUE TITLE
False positive for negated conditions in unless blocks

### ISSUE THREAD

**Issue by woylie** [rrrene/credo#390]
After updating to Credo 0.8.0, I'm getting an `Avoid negated conditions in unless blocks.` However, the referenced function is just a `defmacro __using__(_) do ...` which only contains some aliases, imports, module tags, and functions without any `if` or `unless`.

Credo 0.8.0
Erlang/OTP 19 [erts-8.3] [source-d5c06c6] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

**Comment by NobbZ** [comment]
Are you able to produce a single self contained module or small set of
modules which do provoke that false positive to be better able to find the
problem and also add them as a regression test?

Mathias Polligkeit <notifications@github.com> schrieb am Fr., 2. Juni 2017
um 09:08 Uhr:

**Comment by woylie** [comment]
I just tried to make a mix project with that exact error, but I didn't succeed. In the original project, Credo points directly to the line `defmacro __using__(_) do`. I set up a test project with a module that includes the `__using__` macro, module tags, modules that are imported, modules that are aliased and functions within and outside of the macro body. Nothing caused Credo to emit that warning. So I'm absolutely clueless on what's the cause or how to reproduce the error.

**Comment by NobbZ** [comment]
Can you provide a link to the project that actually causes the error? Or is
it closed source?

Mathias Polligkeit <notifications@github.com> schrieb am Fr., 2. Juni 2017
um 18:25 Uhr:

**Comment by woylie** [comment]
Unfortunately it's closed source.

**Comment by rrrene** [comment]
@woylie Please test against master:

```
{:credo, github: "rrrene/credo"}
```

**Comment by woylie** [comment]
Alas, still the same.

**Comment by rrrene** [comment]
@woylie In that case I am afraid we can't help you without a code sample.

This check hasn't been changed since `v0.1.x`. Here are the cosmetic changes and refactorings to the check from `v0.1.10` to `v0.8.0`:

```shell
diff --git a/lib/credo/check/refactor/negated_conditions_in_unless.ex b/lib/credo/check/refactor/negated_conditions_in_unless.ex
index d80f9bc..d6c867c 100644
--- a/lib/credo/check/refactor/negated_conditions_in_unless.ex
+++ b/lib/credo/check/refactor/negated_conditions_in_unless.ex
@@ -5,13 +5,13 @@ defmodule Credo.Check.Refactor.NegatedConditionsInUnless do
   The code in this example ...
 
       unless !allowed? do
-        proceed_as_planned
+        proceed_as_planned()
       end
 
   ... should be refactored to look like this:
 
       if allowed? do
-        proceed_as_planned
+        proceed_as_planned()
       end
 
   The reason for this is not a technical but a human one. It is pretty difficult
@@ -23,31 +23,31 @@ defmodule Credo.Check.Refactor.NegatedConditionsInUnless do
 
   use Credo.Check, base_priority: :high
 
+  @doc false
   def run(source_file, params \\ []) do
     issue_meta = IssueMeta.for(source_file, params)
 
-    Credo.Code.traverse(source_file, &traverse(&1, &2, issue_meta))
+    Credo.Code.prewalk(source_file, &traverse(&1, &2, issue_meta))
   end
 
   defp traverse({:unless, meta, arguments} = ast, issues, issue_meta) do
-    new_issue =
-      issue_for_first_condition(arguments |> List.first, meta, issue_meta)
+    issue = issue_for_first_condition(List.first(arguments), meta, issue_meta)
 
-    {ast, issues ++ List.wrap(new_issue)}
+    {ast, issues ++ List.wrap(issue)}
   end
   defp traverse(ast, issues, _issue_meta) do
     {ast, issues}
   end
 
   defp issue_for_first_condition({:!, meta, _arguments}, meta, issue_meta) do
-    issue_for(meta[:line], "!", issue_meta)
+    issue_for(issue_meta, meta[:line], "!")
   end
   defp issue_for_first_condition(_, _, _), do: nil
 
 
-  defp issue_for(line_no, trigger, issue_meta) do
+  defp issue_for(issue_meta, line_no, trigger) do
     format_issue issue_meta,
-      message: "Unless conditions should not have a negated condition.",
+      message: "Avoid negated conditions in unless blocks.",
       trigger: trigger,
       line_no: line_no
   end
```

This gives reason to believe that the issue reported for your code should not have been caused by the Credo upgrade alone ...

**Comment by woylie** [comment]
Well, it turns out it was just the wrong message. See [PR #392](https://github.com/rrrene/credo/pull/392)
